lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -404,6 +418,46 @@ void SingleExitLoopTransformer::initialize() {
       llvm::for_each(inst.getResults(), saveEscaping);
     }
   }
+
+  // Do not eliminate undefs unless requested for.
+  if (!TFNoUndefsInSESE)  return;
+
+  // Find a def-free path from the escaping value to the preheader for each
+  // escaping value. If no such path is found, set the escaping value at
+  // preheader to undef.
+  for (auto &kv : escapingValueSubstMap) {
+    const SILValue &escapingValue = kv.first;
+    // FIXME: We should explore the data dependencies in DFS order. Currently,
+    // the order of iteration is arbitrary. DFS should be more efficient.
+    SmallPtrSet<SILValue, 8> worklist;
+    SmallPtrSet<SILValue, 8> visited;
+    worklist.insert(escapingValue);
+    while (!worklist.empty()) {
+      SILValue current = *worklist.begin();
+      visited.insert(current);
+      worklist.erase(current);
+      if (auto *inst = current->getDefiningInstruction()) {
+        if (DI->dominates(inst->getParent(), preheader)) {
+          // Found a definition that we could use.
+          kv.second = current;
+          break;
+        }
+      } else if (auto *arg = dyn_cast<SILArgument>(current)) {
+        if (DI->dominates(arg->getParent(), preheader)) {
+          // Found an argument that we could use.
+          kv.second = current;
+          break;
+        }
+        // This is not usable. Add incoming values to worklist.
+        SmallVector<SILValue, 8> incomingValues;
```

Comment:
the code here suggests if *any* incoming value is in a block that dominates preheader, we can use it.

but actually don't we need *all*, not *any*?

Reply:
Let me rewrite this code using equivalence classes and the idea will be come clear.